File: search_algorithm/python/data_structure/node.py
from problem import Problem
from typing import Optional, List, Any

class Node:
    state: Any
    parent: Optional["Node"]
    action: Optional[Any]
    path_cost: float | int

    def __init__(self, state, parent=None, action=None, path_cost=0):
        self.state = state
        self.parent = parent
        self.action = action
        self.path_cost = path_cost

    def expand(self, problem: Problem):
        children = []
        state = self.state
        for action in problem.actions(state):
            new_state = problem.transition_model(state, action)
            cost = self.path_cost + problem.action_cost(state, action, new_state)
            children.append(Node(new_state, self, action, cost))

        return children

File: search_algorithm/c++/data_structure/node.cc
#include "node.h"

#include <vector>

#include "problem.h"

template <typename TState, typename TAction>
std::vector<std::shared_ptr<Node<TState, TAction>>>
Node<TState, TAction>::Expand(Problem<TState, TAction>& problem) {
    TState current_state = this->get_state();
    std::vector<TAction> actions = problem.GetActions(current_state);
    std::vector<std::shared_ptr<Node<TState, TAction>>> children;

    for (const TAction& action : actions) {
        TState new_state = problem.GetResult(current_state, action);
        float cost = this->get_path_cost() +
                     problem.GetActionCost(current_state, action, new_state);
        children.emplace_back(std::make_shared<Node<TState, TAction>>(
            new_state, shared_from_this(), action, cost));
    }

    return children;
}
